### PR TITLE
Graph Cross Network

### PR DESCRIPTION
**Pull request by paridhimaheshwari2708**
This PR contains the PyG implementation of the paper [Graph Cross Networks with Vertex Infomax Pooling](https://arxiv.org/abs/2010.01804) (NeurIPS 2020). It has been adapted from DGL implementation [here](https://github.com/dmlc/dgl/tree/master/examples/pytorch/gxn), and the author's code is [here](https://github.com/limaosen0/GXN). I have also included an example to show end-to-end training on graph classification task.

Submitted as part of Stanford CS224W: Machine Learning with Graphs, Autumn 2021-2022.

### CODE REVIEW

File: examples/graph_cross_network.py [hunk 1]
```diff
@@ -0,0 +1,248 @@
...
+        return F.log_softmax(logits, dim=1), logits1, logits2
+
+
+def compute_loss(cls_logits: Tensor, labels: Tensor, logits_s1: Tensor,
+                 logits_s2: Tensor, epoch: int, total_epochs: int,
+                 device: torch.device):
+    # classification loss
+    classify_loss = F.nll_loss(cls_logits, labels.to(device))
+
+    # loss for vertex infomax pooling
```
Comment (rusty1s) [inline examples/graph_cross_network.py:108]: Can we integrate the computation of the VIX loss into `GraphCrossNet` model, *e.g.*, `GraphCrossNet.pooling_loss(...)`?

File: examples/graph_cross_network.py [hunk 2]
```diff
@@ -0,0 +1,248 @@
...
+    for batch in trainloader:
+        optimizer.zero_grad()
+        batch.to(device)
+        out, l1, l2 = model(batch)
+        loss = compute_loss(out, batch.y.long(), l1, l2, curr_epoch,
+                            total_epochs, device)
+        loss.backward()
+        optimizer.step()
+
+        total_loss += loss.item()
```
Comment (rusty1s) [inline examples/graph_cross_network.py:140]: Let's compute that similar to [here](https://github.com/pyg-team/pytorch_geometric/blob/master/examples/mutag_gin.py#L76) to account for the last batch size correctly.

File: torch_geometric/nn/models/graph_cross_network.py [hunk 3]
```diff
@@ -0,0 +1,588 @@
...
+import torch.nn.functional as F
+from torch_geometric.data import Batch
+from torch_geometric.utils import subgraph
+from torch_geometric.nn.conv import GCNConv
+from torch_geometric.nn.glob import global_sort_pool
+
+from typing import List, Optional, Union
+
+
+class GCNConvWithDropout(torch.nn.Module):
```
Comment (rusty1s) [inline torch_geometric/nn/models/graph_cross_network.py:13]: Let's not create two much modules here :)

For example, `GCNConvWithDropout`, `DenseLayer` can be safely dropped, IMO.

File: torch_geometric/nn/models/graph_cross_network.py [hunk 4]
```diff
@@ -0,0 +1,588 @@
...
+        """
+        fine_feat = torch.zeros((graph.num_nodes, feat.size(-1)),
+                                device=feat.device)
+        fine_feat[select_idx] = feat
+        fine_feat = self.up_sample_gcn(fine_feat, graph.edge_index)
+        return fine_feat
+
+
+class GraphCrossModule(torch.nn.Module):
+    r"""
```
Comment (rusty1s) [inline torch_geometric/nn/models/graph_cross_network.py:300]: Please follow the docstring conventions of our other modules to ensure that documentation is rendered correctly, see [here](https://github.com/pyg-team/pytorch_geometric/blob/master/torch_geometric/nn/models/mlp.py#L11-L24).

File: torch_geometric/nn/models/graph_cross_network.py [hunk 5]
```diff
@@ -0,0 +1,588 @@
...
+        Default: :obj:`1.0`
+    fuse_weight : float, optional
+        The weight parameter used at the end of GXN for channel fusion.
+        Default: :obj:`1.0`
+    """
+    def __init__(self, pool_ratios: Union[float, List[float]], in_dim: int,
+                 out_dim: int, hidden_dim: int, cross_weight: float = 1.,
+                 fuse_weight: float = 1., dist: int = 1,
+                 num_cross_layers: int = 2):
+        super(GraphCrossModule, self).__init__()
```
Comment (rusty1s) [inline torch_geometric/nn/models/graph_cross_network.py:330]:
```suggestion
        super().__init__()
```

File: torch_geometric/nn/models/graph_cross_network.py [hunk 6]
```diff
@@ -0,0 +1,588 @@
...
+        if isinstance(pool_ratios, float):
+            pool_ratios = (pool_ratios, pool_ratios)
```
Comment (rusty1s) [inline torch_geometric/nn/models/graph_cross_network.py:332]:
```suggestion
            pool_ratios = [pool_ratios] * num_cross_layers
```

File: torch_geometric/nn/models/graph_cross_network.py [hunk 7]
```diff
@@ -0,0 +1,588 @@
...
+    out_dim : int
+        The number of output node feature channels.
+    hidden_dim : int
+        The number of hidden node feature channels.
+    cross_weight : float, optional
+        The weight parameter used in graph cross layers
+        Default: :obj:`1.0`
+    fuse_weight : float, optional
+        The weight parameter used at the end of GXN for channel fusion.
+        Default: :obj:`1.0`
```
Comment (rusty1s) [inline torch_geometric/nn/models/graph_cross_network.py:324]: `num_cross_layers` is missing.
